### PR TITLE
first pass at password grant type

### PR DESCRIPTION
**Pull request by cadecairos**

### CODE REVIEW

File: web/server.js [hunk 1]
```diff
@@ -209,13 +232,25 @@ module.exports = function(options) {
...
             method: function(request, reply) {
               oauthDb.getClient(request.payload.client_id, function(err, client) {
                 if ( err ) {
                   return reply(err);
                 }
-                if ( client.client_secret !== request.payload.client_secret ) {
+                if (
+                  client.allowed_grants.indexOf(request.pre.grant_type) === -1 ||
+                  (
+                    request.pre.grant_type === 'authorization_code' &&
```
Comment (jbuck) [inline web/server.js]: Maybe we should move this logic into the route handler, as opposed to doing it in the pre method

File: web/server.js [hunk 2]
```diff
@@ -225,6 +260,19 @@ module.exports = function(options) {
...
             method: function(request, reply) {
+              if ( request.pre.grant_type === 'password' ) {
+                return account.verifyPassword(request, function(err, json) {
+                  if ( err ) {
+                    return reply(err);
+                  }
+
+                  reply({
+                    user_id: json.user.username,
+                    scopes: request.payload.scopes
```
Comment (jbuck) [inline web/server.js]: This looks like a red flag to me, blindly allowing any scopes. Is there any way we can limit this? Maybe take the scopes from the oauth client whitelist? I can't think of any other option.

Comment (cadecairos) [inline web/server.js]: we could ask for clients to declare scopes in advance for the password grant. I just followed what the spec said to do.

Comment (jbuck) [inline web/server.js]: I know you did, it's just that the spec is really light on details that make the handshake secure :)

File: web/server.js (hunk 1 above)
Comment (cadecairos) [inline web/server.js]: That would require removing all pre methods below into the handler. are we okay with that?

**Comment by cadecairos** [comment]
@jbuck I added some more commits, take a look.

File: web/server.js (hunk 2 above)
Comment (jbuck) [inline web/server.js]: On second thought, if you have the username and password you could set whatever scope you wanted to anyways. Nevermind!

File: web/server.js (hunk 1 above)
Comment (jbuck) [inline web/server.js]: I think we should do it, but lets do it as a follow-up.

File: docs/oauth.md [hunk 3]
```diff
@@ -202,6 +202,62 @@ To log out of id.webmaker.org, use the /logout route
 After a successfully clearing the session, you'll be redirected to the redirect_uri of the supplied client_id, with a `logout=true` in the search portion of the redirect URL
 
 e.g `https://example.org/oauth_redirect?logout=true`
+
+## Native Application Flow
+
+Native applications can use the password grant type to authenticate Webmaker users, without compromising their client_secret.
```
Comment (jbuck) [inline docs/oauth.md]: Should we do something like remove CORS headers on the `/login/oauth/access_token` route to ensure this?

**Comment by jbuck** [comment]
@cadecairos r+

**Pull request merged**